### DISCUSSION TITLE
How can I mutate inputted textField's String with binding action??

### DISCUSSION THREAD

**Discussion by JongHoooon** [pointfreeco/swift-composable-architecture#3062] [Q&A] [answered]
Hellow everyone!

My development eviroment
- TCA: 1.10.3
- iOS Deployment Target: 16.0
- Xcode: 15.3

I want to muate inputted textField's String. ex) lowercased, trimmingCharacters, prefix...

I referenced [Working with SwiftUI bindings](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/bindings/#app-top)

I tried three ways but not works😂

### first try

```Swift
struct ContentView: View {
    @Perception.Bindable private var store: StoreOf<ContentFeature>
    
    init(store: StoreOf<ContentFeature>) {
        self.store = store
    }
    
    var body: some View {
        WithPerceptionTracking {
            VStack {
                TextField("test", text: $store.textFieldText.sending(\.textInpuuted))
            }
            .padding()
        }
    }
}

@Reducer
struct ContentFeature {
    @ObservableState
    struct State: Equatable {
        var textFieldText: String
    }
    
    enum Action: BindableAction {
        case binding(BindingAction<State>)
        case textInpuuted(String)
    }
    
    var body: some ReducerOf<Self> {
        BindingReducer()
        
        Reduce<State, Action> { state, action in
            switch action {
            case let .textInpuuted(text):
                let newText = text.lowercased()
                state.textFieldText = newText
                return .none
                
            default:
                return .none
            }
        }
    }
}
```

### second try

```Swift
struct ContentView: View {
    @Perception.Bindable private var store: StoreOf<ContentFeature>
    
    init(store: StoreOf<ContentFeature>) {
        self.store = store
    }
    
    var body: some View {
        WithPerceptionTracking {
            VStack {
                TextField("test", text: $store.textFieldText)
            }
            .padding()
        }
    }
}

@Reducer
struct ContentFeature {
    @ObservableState
    struct State: Equatable {
        var textFieldText: String
    }
    
    enum Action: BindableAction {
        case binding(BindingAction<State>)
        case textInpuuted(String)
    }
    
    var body: some ReducerOf<Self> {
        BindingReducer()
        Reduce<State, Action> { state, action in
            switch action {
            case .binding(\.textFieldText):
                let newText = state.textFieldText.lowercased()
                state.textFieldText = newText
                return .none
            default:
                return .none
            }
        }
    }
}
```

### thrid try
```Swift
import SwiftUI

import ComposableArchitecture

struct ContentView: View {
    @Perception.Bindable private var store: StoreOf<ContentFeature>
    
    init(store: StoreOf<ContentFeature>) {
        self.store = store
    }
    
    var body: some View {
        WithPerceptionTracking {
            VStack {
                TextField("test", text: $store.textFieldText)
            }
            .padding()
        }
    }
}

@Reducer
struct ContentFeature {
    @ObservableState
    struct State: Equatable {
        var textFieldText: String
    }
    
    enum Action: BindableAction {
        case binding(BindingAction<State>)
        case textInpuuted(String)
    }
    
    var body: some ReducerOf<Self> {
        BindingReducer()
            .onChange(of: \.textFieldText) { oldValue, newValue in
                Reduce<State, Action> { state, action in
                    state.textFieldText = state.textFieldText.lowercased()
                    return .none
                }
            }
        
        Reduce<State, Action> { state, action in
            switch action {
            default:
                return .none
            }
        }
    }
}
```

I also referenced [Performance](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/performance#app-top)

as I read sending actions is not lightweight so mutating inputted string every time looks not good.

Is there any good way mutating inpuuted string??

**Comment by mbrandonw** [comment]
Hi @JongHoooon, this is just a limitation of SwiftUI in general. If you were to recreate this app using either `@State` or `@Observable` you would see the same behavior. For example, this:

```swift
@Observable
class Model {
  var text = "hello" {
    didSet {
      if text != text.lowercased() {
        text = dump(text.lowercased())
      }
    }
  }
}

struct A: View {
  @Bindable var model = Model()
  var body: some View {
    Form {
      TextField("", text: $model.text)
    }
  }
}
```

I believe the way SwiftUI wants you to do these kinds of transformations is through formatters, which is an additional argument you can pass to `TextField`.

**Reply by JongHoooon** [reply]
Thank you @mbrandonw 

As I know Using formatter works when keyboard return button tapped. I want to check and mutate everytime user input string. 

I also tried not using action. It works well. Is this related to performance(action is not lightweight ) ??
 
```Swift
var body: some View {
    WithPerceptionTracking {
        VStack {
            TextField("test", text: $store.textFieldText)
                .onChange(of: store.textFieldText) { newValue in
                    store.textFieldText = store.textFieldText.lowercased()
                }
        }
        .padding()
    }
}
```

**Reply by mbrandonw** [reply, marked as answer]
That trick seems good to me!

**Reply by sbilling** [reply]
This solution does not work anymore as attempting to reference the binding directly on the textfield throws `Cannot assign through dynamic lookup property: subscript is get-only`. 

@mbrandonw any thoughts here?

**Reply by mbrandonw** [reply]
@sbilling this works fine as long as you have set up a binding reducer to unlock the `store.field = value` syntax.

**Reply by freak4pc** [reply]
> Thank you @mbrandonw
> 
> As I know Using formatter works when keyboard return button tapped. I want to check and mutate everytime user input string.
> 
> I also tried not using action. It works well. Is this related to performance(action is not lightweight ) ??
> 
> ```swift
> var body: some View {
>     WithPerceptionTracking {
>         VStack {
>             TextField("test", text: $store.textFieldText)
>                 .onChange(of: store.textFieldText) { newValue in
>                     store.textFieldText = store.textFieldText.lowercased()
>                 }
>         }
>         .padding()
>     }
> }
> ```

Sorry for bumping this one - 
I'm a bit confused with this being the recommended approach. Isn't this exactly what TCA tries to go **aginst**? i.e. moving actual business logic into the view layer this way seems extremely untestable, as opposed to being able to do this in the reducer? 🤔

**Reply by freak4pc** [reply]
(yes i'm aware this is going through the reducer, but this logic is basically invoked in the view, it's another place a developer has to look at)

**Reply by mbrandonw** [reply]
> I'm a bit confused with this being the recommended approach. Isn't this exactly what TCA tries to go **aginst**? i.e. moving actual business logic into the view layer this way seems extremely untestable, as opposed to being able to do this in the reducer? 🤔

This is ultimately a SwiftUI bug that even affects vanilla SwiftUI apps (any app using `@Observable` models). So I'm not sure what else we can do other than give tips on how to help people move forward even if it may not be ideal.

And while a goal of TCA is to move as much logic as possible to the reducer so that it is testable, that is not always feasible. We have to sometimes face reality and just cut our loses.

If you have better ideas for this problem then please do share so that others can benefit.

**Reply by freak4pc** [reply]
I might be mistaken but doesn't a standard action-based binding solve this? (i.e. not using BindingAction and synthesized bindings) ? 

Because you could have your custom action verify/validate/mutate the text in any way it sees feet when the custom binding fires the action. (Or I'm missing something?)

**Reply by mbrandonw** [reply]
> I might be mistaken but doesn't a standard action-based binding solve this? (i.e. not using BindingAction and synthesized bindings) ?

No, even if you have a dedicated `textChanged` action that does the following in the reducer:

```swift
case .textChanged(let text):
  state.text = text.uppercased()
```

…and then use it in the view like this:

```swift
TextField("", text: $store.text.sending(\.textChanged)
```

…the text will _not_ be uppercased while typing into the text field. And the exact same thing happens in an observable model:

```swift
@Observable
class Model {
  var text = "hello" {
    didSet {
      if text != text.uppercased() {
        text = text.uppercased()
      }
    }
  }
}
```

…and then use that in a view:

```swift
TextField("", text: $model.text)
```

The state in the "source of true" and the state visible in the UI will drift apart. So this definitely is just a SwiftUI bug. 

The only time these kinds seem to work is when using `@State`, but then that is fraught because tapping into the `didSet`/`willSet` of property wrappers has very nuanced rules of when it is allowed and when it will silently ignore the callbacks.

I suppose another option is to leverage `@State` as a kind of intermediary between your feature and the text field:

```swift
struct FeatureView: View {
  @Bindable var store: StoreOf<Feature>
  @State var text = ""
  var body: some View {
    Form {
      TextField("", text: $text)
    }
    .onChange(of: store.text) { text = store.text }
    .onChange(of: text) { store.send(.textChanged(text) }
  }
}
```

You would probably need to do equality checks in the `onChange` in order to prevent an infinite loop. You could even make use of the `bind` helper we have in SwiftUINavigation:

```swift
struct FeatureView: View {
  @Bindable var store: StoreOf<Feature>
  @State var text = ""
  var body: some View {
    Form {
      TextField("", text: $text)
    }
    .bind($store.text.send(\.textChanged), to: $text)
  }
}
```

I haven't tested it, but I believe that might work around the SwiftUI bug and keep the logic in the TCA reducer.

**Comment by alvarezGarciaMarcos** [comment]
I've found the same problem. In my case I wanted to limit the length of the String that is typed by the user. I need to modify the length of the String so that when the user erases one character, the character is deleted from the limited-length String. Did you find a solution to this?

**Reply by mbrandonw** [reply]
Hi @alvarezGarciaMarcos, there is a solution from @JongHoooon in the discussion [above](#discussioncomment-9377436).

**Comment by JongHyunLee84** [comment]
@JongHoooon @alvarezGarciaMarcos @mbrandonw 
I've discovered something interesting but don't know why.... 
Using TextEditor instead of TextField works well in [this second case](https://github.com/pointfreeco/swift-composable-architecture/discussions/3062#discussion-6638102)
This allows you to create a custom TextEditor similar to TextField
Personally, I believe that mutating state outside of the reducer, as shown in [this example](https://github.com/pointfreeco/swift-composable-architecture/discussions/3062#discussioncomment-9377436), doesn't align with the TCA style

**Reply by mbrandonw** [reply]
> Personally, I believe that mutating state outside of the reducer, as shown in https://github.com/pointfreeco/swift-composable-architecture/discussions/3062#discussioncomment-9377436, doesn't align with the TCA style

For what it's worth, this does _not_ mutate state outside of the reducer. Doing `store.textFieldText = …` is actually sending the `binding` action to the reducer in order to update the state. A `Store`'s state is always [read-only](https://github.com/pointfreeco/swift-composable-architecture/blob/8343030d8d9891406c0c4bb9a298618ec0e30dd2/Sources/ComposableArchitecture/Observation/Store%2BObservation.swift#L18-L25), but when binding actions are available you get to user the shortened syntax `store.field = …` instead of `store.send(.binding(\.field, …))`.

**Reply by JongHyunLee84** [reply]
I thought it was directly modifying the state, but it wasn't. Thank you for letting me know.

**Reply by JongHoooon** [reply]
@mbrandonw @JongHyunLee84 

I know it was not directly modifying the state. but I think view had business logic looks not good. I want to handling every business logic in the reducer.
